File: day-04/python/part1.py
#!/usr/bin/env python3

import sys

required_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]  # "cid" not needed as it is optional

passports = [f for f in open(sys.argv[1], 'r').read().split("\n\n")]

valid = 0

for field in passports:
    d = {}
    for b in field.split():
        k, v = b.split(":")
        d[k] = v

    invalid = False
    for req in required_fields:
        if req not in d.keys():
            invalid = True

    if not invalid:
        valid += 1

print(valid)

File: day-04/nim/part1.nim
from os import paramStr
import tables
import strutils
from sequtils import toSeq

let
  required_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
  passports = paramStr(1).string.readfile.split("\n\n")

var valid: int = 0

for field in passports:
  var d = initTable[string, string]()
  for b in field.split():
    var kv = b.split(":")
    d[kv[0]] = kv[1]

  var invalid = false
  for req in required_fields:
    if req notin toSeq(d.keys):
      invalid = true

  if not invalid:
    valid += 1

echo valid
